### ISSUE TITLE
Deprecate Docker socket over TCP without TLS

### ISSUE THREAD

**Issue by justincormack** [moby/moby#36357]
Running the Docker socket over TCP without any authentication, even on `localhost` (yes, JavaScript can connect to localhost from your browser) is very dangerous, even in non production environments. Since we have had named pipe support for Windows for some years now, and always have had unix socket support, there is no reason for this to be supported out of the box, and we should remove it. Obviously we will continue to support tcp with TLS.

If users really want to do this, they can make a proxy from tcp to the socket, which is pretty trivial, but then it is clearly their responsibility.

We could potentially have quite a long deprecation period, and should have warnings on the client and server.

**Comment by AkihiroSuda** [comment]
SGTM but maybe except DIND?

**Comment by justincormack** [comment]
@AkihiroSuda why do you need it in dind?

**Comment by AkihiroSuda** [comment]
a nondind container may want to connect to a dind container.
I think non-tls is allowed when separate nw is created.

**Comment by justincormack** [comment]
The problem is that Docker cannot know if there is a separate network. We could have a flag to allow the current behaviour still, but I would still rather you had to run a proxy. The dind container could include a suitable proxy option I guess.

**Comment by justincormack** [comment]
We could add a feature to generate a TLS keypair for simple local use cases.

**Comment by AkihiroSuda** [comment]
Yes, I think just including socat is fine (after some deprecation period of the current behaviour)

**Comment by AkihiroSuda** [comment]
Still SGTM

**Comment by cyli** [comment]
It sounds like from the description the comments in from the comments from #37299, this feature for deprecating support for tcp without *mutual* TLS, as opposed to unauthenticated one-way TLS.  If that's the case, would it make sense to clarify on the issue title?

**Comment by corhere** [comment]
https://github.com/moby/moby/pull/41285#issuecomment-694422616
> - future release; enable both `--tls` and `--tlsverify` by default: if a user didn't specify options, the daemon will fail to start as no certs are provided
> - add options to explicitly disable tls / tlfverify: `--disable-tls`, `--disable-tls-verify` (name can be discussed)

**Comment by corhere** [comment]
An entry needs to be added to the [deprecation log](https://github.com/docker/cli/blob/master/docs/deprecated.md) before we can enable `--tlsverify` by default.

**Comment by neersighted** [comment]
Now that we're getting serious about doing this, paging @tianon for thoughts on what to do for `docker:dind`. Whatever we do there, we'll probably want to replicate for our dev container (I make use of the TCP API a fair bit for testing, at least).

**Comment by tianon** [comment]
We've had the `docker:dind` image doing automatic generation of all the necessary self-signed certificates for a while now, with honestly pretty limited success.

(For details, see the insanely hairy mess that is https://github.com/docker-library/docker/blob/1b4d76a2e8049074e723b5b8ecec417b1450a065/dockerd-entrypoint.sh -- generating the certs correctly/acceptably/securely in the first place is way harder than it ought to be IMO, but we also have to be careful not to clobber certificates a user might provide)

One of the biggest sets of users that complains any time anything breaks in that image is GitLab, and last I knew it was kind of complicated for them to set up certificates correctly for this to work for their pipelines.  I don't know whether that's changed, but I suspect it's _still_ complicated because they (GitLab) don't directly control how users include this in their YAML, and many of their users have copy-pasted some boilerplate, and that's the main source of the consternation any time anything changes (like the recent `xtables` vs `nftables` snafu).

See https://github.com/docker-library/docker/pull/430 for a currently-active example of a contributor/user who is pretty frustrated about this.

IMO, turning the current delay into an error is OK if we have an escape hatch for users to say "yes, I accept the risk, please be really insecure", but I don't personally think `--tls=false --tlsverify=false` really conveys the severity of how insecure this configuration is appropriately. :disappointed:

**Comment by trallnag** [comment]
In any case you should be able to use something like socat to proxy the unix socket via a tcp socket. Like stated in the initial post

**Comment by brwilkinson** [comment]
@justincormack I am running Moby on Windows.

Using IPSEC tunnel between machines. This is ideal, since we don't need to manage certs or keys. We utilize Kerberos + Windows Firewall. 

Please allow a flag in the docker config 'daemon.json' so that I can continue to do.

```pwsh
Get-NetIPsecRule


IPsecRuleName         : Docker-IPSEC-TCP-2375
DisplayName           : Docker-IPSEC-TCP-2375
Description           :
DisplayGroup          :
Group                 :
Enabled               : True
Profile               : Any
Platform              : {}
Mode                  : Transport
InboundSecurity       : Request
OutboundSecurity      : Request
QuickModeCryptoSet    : Default
Phase1AuthSet         : ComputerKerberos
Phase2AuthSet         : UserKerberos
KeyModule             : Default
AllowWatchKey         : False
AllowSetKey           : False
LocalTunnelEndpoint   :
RemoteTunnelEndpoint  :
RemoteTunnelHostname  :
ForwardPathLifetime   : 0
EncryptedTunnelBypass : False
RequireAuthorization  : False
User                  : Any
Machine               : Any
PrimaryStatus         : OK
Status                : The rule was parsed successfully from the store. (65536)
EnforcementStatus     : NotApplicable
PolicyStoreSource     : PersistentStore
PolicyStoreSourceType : Local

Get-NetIPsecMainModeSA


Name                                : 67
LocalEndpoint                       : 10.19.10.163
RemoteEndpoint                      : 10.19.10.167
LocalFirstId.Identity               : host/myclient.domain.com
LocalFirstId.ImpersonationType      : None
LocalFirstId.AuthenticationMethod   : Kerberos
LocalFirstId.Flags                  : None
RemoteFirstId.Identity              : host/mydockerhost.domain.com
RemoteFirstId.ImpersonationType     : None
RemoteFirstId.AuthenticationMethod  : Kerberos
RemoteFirstId.Flags                 : None
LocalSecondId.Identity              : domain\svcgmsaaccount$
LocalSecondId.ImpersonationType     : Impersonated
LocalSecondId.AuthenticationMethod  : Kerberos
LocalSecondId.Flags                 : None
RemoteSecondId.Identity             : host/mydockerhost.domain.com
RemoteSecondId.ImpersonationType    : None
RemoteSecondId.AuthenticationMethod : Kerberos
RemoteSecondId.Flags                : None
CipherAlgorithm                     : AES128
HashAlgorithm                       : SHA1
GroupId                             : None
KeyModule                           : AuthIP
MaxQMSAs                            : 0
LifetimeSeconds                     : 28800
LocalUdpEncapsulationPort           :
RemoteUdpEncapsulationPort          :
ExtendedFilterId                    : 0
```

**Comment by brwilkinson** [comment]
@justincormack I wanted to follow up. 
- What are the chances of a flag so we can continue to leverage the Windows built in Firewall, with IPSEC rules, that allow only configured clients/users to connect, via the IPSEC tunnel?

**Comment by brwilkinson** [comment]
I just saw this was opened in 2018... I would like consideration for an override based on my IPSEC method I use, mentioned in the previous comments. Otherwise, could we just close this out? Sure, running without security is not smart, however the warning we have provides people with sufficient advice on what to do next. Given there are other ways to run securely, i would prefer if we can just get this marked as closed as will not fix.
